### PR TITLE
Client Authentication

### PR DESCRIPTION
**Pull request by MUzzell**
Hello! 

I need to find a means of getting a Windows Service to push its logs to a rsyslog server securely. I use NLog already and thought this project would be brilliant to achieve this.
However, to get rsyslog to work with TLS, the client (my service) needs to be able to authenticate with the server by providing a client certificate. 

To that end, I have some minor additions to the `Tcp` process to allow Client certs to be selected from the Windows Cert stores. A user can choose from one of the Store location enums (`StoreName` and `StoreLocation`) and also find a cert based on subject name.

### CODE REVIEW

File: src/NLog.Targets.Syslog/MessageSend/Tcp.cs [hunk 1]
```diff
@@ -66,7 +67,7 @@ private bool IsSocketConnected()
             if (connectionCheckTimeout <= 0)
                 return true;
 
-            return  tcp.Client.Poll(connectionCheckTimeout, SelectMode.SelectRead) && tcp.Client.Available == 0;
+            return  tcp.Client.Poll(connectionCheckTimeout, SelectMode.SelectWrite) && tcp.Client.Available == 0;
```
Comment (luigiberrettini) [inline src/NLog.Targets.Syslog/MessageSend/Tcp.cs]: This change is not related with this PR (it should have its own PR).
Anyway SelectRead should be used:
     - https://msdn.microsoft.com/en-us/library/system.net.sockets.socket.poll.aspx
     - http://stackoverflow.com/questions/722240/instantly-detect-client-disconnection-from-server-socket
     - http://blog.stephencleary.com/2009/05/detection-of-half-open-dropped.html

File: src/NLog.Targets.Syslog/MessageSend/Tcp.cs [hunk 2]
```diff
@@ -98,9 +111,23 @@ private Stream SslDecorate(TcpClient tcpClient)
...
                 return tcpStream;
 
+            X509Certificate2Collection clientCerts = null;
+            if (clientCertStore != null)
+            {
+                clientCertStore.Open(OpenFlags.ReadOnly);
+                clientCerts = clientCertStore.Certificates;
+
+                if (clientCertSubjectName != null)
+                    clientCerts = clientCerts.Find(X509FindType.FindBySubjectName, clientCertSubjectName, false);
```
Comment (luigiberrettini) [inline src/NLog.Targets.Syslog/MessageSend/Tcp.cs]: What if we move the `clientCertStore.Close();` here?
Can you test if it works?

File: src/NLog.Targets.Syslog/MessageSend/Tcp.cs (hunk 1 above)
Comment (MUzzell) [inline src/NLog.Targets.Syslog/MessageSend/Tcp.cs]: Thanks for looking at my PR, the blog in particular is very helpful!

I made this change to because the socket was being torn down and setup for every message as `IsSocketConnected` was always returning false. I decided to change it on this PR because on my rsyslog server it was causing the following and felt it was important to change:
```
Mar 12 21:33:03 DESKTOP-3JEJ4K2 TestApp.FormTest[1784] ﻿This is a sample info message
Mar 12 21:33:16 muzzell-rsyslog rsyslogd-2078: unexpected GnuTLS error -110 in nsd_gtls.c:530: The TLS connection was non-properly terminated.  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:16 muzzell-rsyslog rsyslogd-2078: netstream session 0x7fdd780121d0 from 10.8.0.22 will be closed due to error  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:05 DESKTOP-3JEJ4K2 TestApp.FormTest[1784] ﻿This is a sample info message
Mar 12 21:33:18 muzzell-rsyslog rsyslogd-2078: unexpected GnuTLS error -110 in nsd_gtls.c:530: The TLS connection was non-properly terminated.  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:18 muzzell-rsyslog rsyslogd-2078: netstream session 0x7fdd78011c40 from 10.8.0.22 will be closed due to error  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:08 DESKTOP-3JEJ4K2 TestApp.FormTest[1784] ﻿This is a sample info message
Mar 12 21:33:21 muzzell-rsyslog rsyslogd-2078: unexpected GnuTLS error -110 in nsd_gtls.c:530: The TLS connection was non-properly terminated.  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:21 muzzell-rsyslog rsyslogd-2078: netstream session 0x7fdd78011de0 from 10.8.0.22 will be closed due to error  [v8.16.0 try http://www.rsyslog.com/e/2078 ]
Mar 12 21:33:10 DESKTOP-3JEJ4K2 TestApp.FormTest[1784] ﻿This is a sample info message
```

Is that tear-down intentional, or do you think is my SSL setup getting in the way somehow?

**Comment by luigiberrettini** [comment]
I put client certificate configuration under a new section that also includes TLS.
Please @MUzzell take a look at the code and give me feedback about readability and naming besides telling me if everything works as expected.

**Comment by MUzzell** [comment]
I don't have time to test tonight but ill give it a bash tomorrow. Looking over the changes I can't see anything wrong. 
One thing I guess would be to update the README (which I'm happy to do tomorrow).
Thanks again for the review!

**Comment by MUzzell** [comment]
I have tested the search types, all looks good.
I have also updated the readme with the new tls params.

**Comment by luigiberrettini** [comment]
Since this PR introduces major changes to configuration a new release is needed:
merge will be performed after fixing #82.

**Comment by MUzzell** [comment]
Awesome! Thanks so much for looking at this, hope I did not cause too much trouble for you 😄.

**Comment by luigiberrettini** [comment]
Closes #84
